librz/type/path.c

```diff
@@ -36,74 +36,160 @@ RZ_API void rz_type_path_free(RZ_NULLABLE RzTypePath *tpath) {
+	size_t nd = 0;
+	RzType *typd = parent;
+	st64 curd_off = 1; // prod of all dim sizes
+	while (typd->kind == RZ_TYPE_KIND_ARRAY) {
+		++nd;
+		curd_off *= typd->array.count; // TODO: overflow
+
+		typd = typd->array.type;
+	}
+	curd_off *= rz_type_get_base_type(typedb, typd)->size; // elem size in bits
+
+	typd = parent;
+	for (size_t id = 0; id < nd; ++id) {
+
+		if (path[*i] != '[') {
+			eprintf("Expected '[' got '%c'\n", path[*i]);
+			*offset = -1;
+			return NULL;
+		}
+
+		++*i;
+		size_t tok_beg = *i;
+		for (; isdigit(path[*i]); ++*i)
+			;
+
+		if (path[*i] != ']') {
+			eprintf("Expected ']' got '%c'\n", path[*i]);
+			*offset = -1;
+			return NULL;
+		}
+
+		size_t idx = strtoull(strndup(&path[tok_beg], *i - tok_beg), NULL, 10);
+		curd_off /= typd->array.count;
+
+		*offset += curd_off * idx;
+		typd = typd->array.type;
+
+		// DONT check for "idx < array size"
+		// because guy may be exploiting OOB or something else.
+		// so just continue.
+
+		++*i; // skip ']'
+	}
+
+	return typd;
+}
+
+static RzType *path_walker_parse_dot(const RzTypeDB *typedb, RzType *parent, const char *path, size_t *i, st64 *offset) {
+	++*i;
+
+	size_t tok_beg = *i;
+	for (; isalnum(path[*i]); ++*i)
+		;
+
+	const char *tok = strndup(&path[tok_beg], *i - tok_beg);
+
+	RzBaseType *parent_btype = rz_type_get_base_type(typedb, parent);
+	if (!parent_btype) {
+		eprintf("Could not found btype for parent\n");
+		*offset = -1;
+		return NULL;
+	}
+
+	RzTypeStructMember *memb_it;
+	RzType *cur_type = NULL;
+	size_t cur_offset = -1;
+	rz_vector_foreach(&parent_btype->struct_data.members, memb_it) {
+
+		if (!strcmp(memb_it->name, tok)) {
+			cur_type = memb_it->type;
+
+			cur_offset = memb_it->offset; // in bytes
+			cur_offset *= 8; // in bits
+			break;
+		}
+	}
+
+	if (!cur_type) {
+		eprintf("Invalid member '%s' for parent type\n", tok);
+		*offset = -1;
+		return NULL;
+	}
+
+	*offset += cur_offset;
+
+	if (path[*i] == '[') {
+
+		parent = cur_type;
+
+		if (parent->kind != RZ_TYPE_KIND_ARRAY) {
+			eprintf("Expected array, got another type\n");
+			*offset = -1;
+			return NULL;
+		}
+
+		path_walker_parse_bracket(typedb, parent, path, i, offset);
+		if (*offset == -1) {
+			return NULL;
+		}
+
+		// unwrap array
+		while (parent->kind == RZ_TYPE_KIND_ARRAY) {
+			parent = parent->array.type;
+		}
+
+		return parent;
+	} else if (path[*i] == '.' || path[*i] == '\0') {
+		return cur_type;
+	} else {
+		eprintf("Unexpected character '%c' at position %lu\n", path[*i], *i);
+		*offset = -1;
+		return NULL;
+	}
+}
+
 static st64 path_walker(const RzTypeDB *typedb, const char *path) {
 	rz_return_val_if_fail(typedb && path, -1);
-	const char *member;
-	size_t membsize;
-	ut64 index;
+
+	size_t i;
+	for (i = 0; isalnum(path[i]); ++i)
+		;
+
+	if (path[i] != '.') {
+		eprintf("Unexpected character '%c' at position %lu\n", path[i], i);
+		return -1;
+	}
+
+	RzType *parent = rz_type_identifier_of_base_type_str(typedb, strndup(path, i));
```

Comment:
Should the strndup string be assigned to a var and freed later?

Reply:
Fixed here, thanks